File: flink/src/main/java/com/zoo/flink/java/sink/SinkToMysqlDemo.java
package com.zoo.flink.java.sink;

import com.zoo.flink.java.FlinkEnv;
import org.apache.flink.connector.jdbc.JdbcConnectionOptions;
import org.apache.flink.connector.jdbc.JdbcExecutionOptions;
import org.apache.flink.connector.jdbc.JdbcSink;

/**
 * @Author: JMD
 * @Date: 5/11/2023
 */
public class SinkToMysqlDemo extends FlinkEnv {
    // 对于 MySQL 5.7， 用"com.mysql.jdbc.Driver"
    public static void main(String[] args) throws Exception {
        arrayStream.addSink(
                JdbcSink.sink(
                        "INSERT INTO clicks (user, url) VALUES (?, ?)",
                        (statement, r) -> {
                            statement.setString(1, r.user);
                            statement.setString(2, r.url);
                        },
                        JdbcExecutionOptions.builder()
                                .withBatchSize(1000)
                                .withBatchIntervalMs(200)
                                .withMaxRetries(5)
                                .build(),
                        new JdbcConnectionOptions.JdbcConnectionOptionsBuilder()
                                .withUrl("jdbc:mysql://hadoop102:3306/data")
                                .withDriverName("com.mysql.cj.jdbc.Driver")
                                .withUsername("root")
                                .withPassword("1234")
                                .build()
                ));
        env.execute();
    }
}

File: flink/src/main/scala/com/zoo/flink/scala/sink/SinkToMysqlDemo.scala
package com.zoo.flink.scala.sink

import com.zoo.flink.java.util.Event
import com.zoo.flink.scala.util.FlinkEnv
import org.apache.flink.connector.jdbc.{JdbcConnectionOptions, JdbcExecutionOptions, JdbcSink, JdbcStatementBuilder}

import java.sql.PreparedStatement

/**
 * Author: JMD
 * Date: 5/12/2023
 */
object SinkToMysqlDemo extends FlinkEnv{
  def main(args: Array[String]): Unit = {
    val jdbcStatementBuilder: JdbcStatementBuilder[Event] = new JdbcStatementBuilder[Event](){
      override def accept(ps: PreparedStatement, event: Event): Unit = {
        ps.setString(1, event.user)
        ps.setString(2, event.url)
      }
    }

    // 用函数方式传递 jdbcStatementBuilder就报错类型不匹配，没能解决，现在只可以用匿名类
    arrayStream.addSink(
      JdbcSink.sink(
        s"INSERT INTO clicks (user, url) VALUES (?, ?)",
/*        (ps: PreparedStatement, event: Event) => {
          ps.setString(1, event.user)
          ps.setString(2, event.url)
        }.asInstanceOf[JdbcStatementBuilder[Event]],*/
        jdbcStatementBuilder,
        JdbcExecutionOptions.builder()
          .withBatchSize(1000)
          .withBatchIntervalMs(200)
          .withMaxRetries(5)
          .build(),
        new JdbcConnectionOptions.JdbcConnectionOptionsBuilder()
          .withUrl("jdbc:mysql://hadoop102:3306/data")
          .withDriverName("com.mysql.cj.jdbc.Driver")
          .withUsername("root")
          .withPassword("1234")
          .build()
      ))

    env.execute
  }
}
